### ISSUE TITLE
Kinematic fit over constrained in fitted pi0 channels?

### ISSUE THREAD

**Issue by acschick** [JeffersonLab/halld_recon#733]
I'm looking at events from the gp->pi0(p) and gp->pi0 pi0 (p) reaction filters. My intuitions tell me that the kin fit momenta of the two photons summed together should match the DecayingPi0 kin fit momentum--instead I see a large discrepancy. This occurs in both the single pi0 and double pi0 channels.

<img width="714" alt="Screenshot 2023-05-12 at 3 17 09 PM" src="https://github.com/JeffersonLab/halld_recon/assets/8531932/28bfe34b-89d6-46b0-b711-89355b208945">

[Invariant Mass Plots.pdf](https://github.com/JeffersonLab/halld_recon/files/11480772/Invariant.Mass.Plots.pdf)

Is this a bug? I'm just using the auto-generated DSelector code to pull out the momentum vectors and histogram them.
Example DSelector here: /work/halld/home/shong/channels/pi0missprot/BasicRunScripts/DSelector_pi0missprot.C

**Comment by zihlmann** [comment]
Using Andrews code and running over 10 files then look at the two histogram 
DecayingPi0_W and 2Photon_W do not compute because the DecayingPi0_W
is the mass constraint Pi0, hence a delta function, while the 2Photon_W is a distribution
which should not be because it is made from the two photons from the kin fit.
![decayingPI0](https://github.com/JeffersonLab/halld_recon/assets/13365259/0c53005b-d27f-4615-b173-b9408609ef44)

![2gamma_invMass_KinFIT](https://github.com/JeffersonLab/halld_recon/assets/13365259/0041f9e5-e28a-4b7c-a306-cf3ee1210a3a)

**Comment by jonzarling** [comment]
Definitely looks wonky. Maybe the interplay between mass constrained pi0 and missing proton is some edge case we haven't encountered before? Like maybe it forgets that a missing proton can still be mass constrained, and this somehow leaks into the calculation of gamma1 and gamma2. Maybe checking the hd_root.root folders with kinfit info or some sanity checks calculating the same quantity different ways will turn something up....

**Comment by zihlmann** [comment]
The missing Mass looks like the plot below requiring events with exactly 2 reconstructed photons. the red histograms
is after cutting 2 sigma around the pi0 invariant mass.
![missing_mass_with_andwithoutcut](https://github.com/JeffersonLab/halld_recon/assets/13365259/170105dc-93f8-4663-838b-ba80bfbdf006)

**Comment by zihlmann** [comment]
Yes, indeed, I also think that the code somehow does not contrain the pi0 mass the same way as usual when there is a missing proton involved. In the plot below the red histogram represents the measured values while the blue the kinfit values. there are only small changes best seen at the "edges".
![pi0invmass_kinfit_measuredinred](https://github.com/JeffersonLab/halld_recon/assets/13365259/41072140-1484-44a3-a84e-36f312cd212b)

**Comment by sdobbs** [comment]
I'll see if I have some time to look into this in the near future, but let me suggest you redo the fits with the "F1" option, which only imposes the 4-momentum constraints, and removes the vertex constraint from the fit - since there's no charged particle in the final state, there's no possible measurement of the vertex position, so that could be confounding the fit.

As a note, for the gp -> pi0 (p) fit, this should then be a 2C fit - one constraint for the pi0 mass in pi0 -> gg, and 1 constraint for the missing mass in the reaction being consistent with that of the proton.  If the F1 fits still show a problem, the next step would be to look at these constraints and make sure they are not beating against each other in some strange way.
